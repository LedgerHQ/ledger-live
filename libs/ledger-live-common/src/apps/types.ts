import { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
import type { DeviceModel, DeviceModelId } from "@ledgerhq/devices";
import { App, DeviceInfo, FinalFirmware } from "@ledgerhq/types-live";
import type { Observable, Subject } from "rxjs";
export type Exec = (
  appOp: AppOp,
  targetId: string | number,
  app: App
) => Observable<{
  progress: number;
}>;
export type InstalledItem = {
  name: string;
  updated: boolean;
  hash: string;
  blocks: number;
  version: string;
  availableVersion: string;
};
export type ListAppsEvent =
  | {
      type: "device-permission-requested";
      wording: string;
    }
  | {
      type: "device-permission-granted";
    }
  | {
      type: "result";
      result: ListAppsResult;
    }
  | {
      type: "allow-manager-requested";
    };
export type ListAppsResult = {
  appByName: Record<string, App>;
  appsListNames: string[];
  installedAvailable: boolean;
  installed: InstalledItem[];
  deviceInfo: DeviceInfo;
  deviceModelId: DeviceModelId;
  firmware: FinalFirmware | null | undefined;
};
export type State = {
  deviceInfo: DeviceInfo;
  deviceModel: DeviceModel;
  firmware: FinalFirmware | null | undefined;
  appByName: Record<string, App>;
  apps: App[];
  installedAvailable: boolean;
  installed: InstalledItem[];
  recentlyInstalledApps: string[];
  installQueue: string[];
  uninstallQueue: string[];
  // queue saved at the time of a "updateAll" action
  updateAllQueue: string[];
  currentAppOp: AppOp | null | undefined;
  currentProgressSubject: Subject<number> | null | undefined;
  currentError:
    | {
        error: Error;
        appOp: AppOp;
      }
    | null
    | undefined;
};
export type AppOp =
  | {
      type: "install";
      name: string;
    }
  | {
      type: "uninstall";
      name: string;
    };
export type Action =  // recover from an error
  | {
      type: "recover";
    } // wipe will remove all apps of the device
  | {
      type: "wipe";
    } // uninstall a specific app by name
  | {
      type: "uninstall";
      name: string;
      force?: boolean;
    } // install or update a specific app by name
  | {
      type: "install";
      name: string;
    } // update all
  | {
      type: "updateAll";
    } // action to run after an update was done on the device (uninstall/install)
  | {
      type: "onRunnerEvent";
      event: RunnerEvent;
    };
export type RunnerEvent =
  | {
      type: "runStart";
      appOp: AppOp;
    }
  | {
      type: "runProgress";
      appOp: AppOp;
      progress: number;
    }
  | {
      type: "runError";
      appOp: AppOp;
      error: Error;
    }
  | {
      type: "runSuccess";
      appOp: AppOp;
    };
export type AppData = {
  currency: CryptoCurrency | null | undefined;
  name: string;
  blocks: number;
  bytes: number;
};
export type UnrecognizedAppData = {
  name: string;
  hash: string;
};
export type AppsDistribution = {
  totalBlocks: number;
  totalBytes: number;
  osBlocks: number;
  osBytes: number;
  apps: Array<AppData>;
  appsSpaceBlocks: number;
  appsSpaceBytes: number;
  totalAppsBlocks: number;
  totalAppsBytes: number;
  freeSpaceBlocks: number;
  freeSpaceBytes: number;
  shouldWarnMemory: boolean;
};
