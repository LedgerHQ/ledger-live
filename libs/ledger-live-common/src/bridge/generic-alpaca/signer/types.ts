export type SignTransactionOptions = {
  rawTxHex: string;
  path: string;
};
