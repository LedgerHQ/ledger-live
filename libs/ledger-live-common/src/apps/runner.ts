import { Observable, from, of, defer, concat } from "rxjs";
import {
  map,
  materialize,
  reduce,
  ignoreElements,
  throttleTime,
  scan,
  mergeMap,
  distinctUntilChanged,
} from "rxjs/operators";
import type { Exec, State, AppOp, RunnerEvent, Action } from "./types";
import { reducer, getActionPlan, getNextAppOp } from "./logic";
import { delay } from "../promise";
import { getEnv } from "../env";

export const runAppOp = (
  { appByName, deviceInfo }: State,
  appOp: AppOp,
  exec: Exec
): Observable<RunnerEvent> => {
  const app = appByName[appOp.name];

  if (!app) {
    const events: RunnerEvent[] = [
      {
        type: "runStart",
        appOp,
      },
      {
        type: "runSuccess",
        appOp,
      },
    ];
    // app not in list, we skip it.
    return from(events);
  }

  return concat(
    of(<RunnerEvent>{
      type: "runStart",
      appOp,
    }), // we need to allow a 1s delay for the action to be achieved without glitch (bug in old firmware when you do things too closely)
    defer(() => delay(getEnv("MANAGER_INSTALL_DELAY"))).pipe(ignoreElements()),
    defer(() => exec(appOp, deviceInfo.targetId, app)).pipe(
      throttleTime(100),
      materialize(),
      map((n) => {
        switch (n.kind) {
          case "N":
            return <RunnerEvent>{
              type: "runProgress",
              appOp,
              progress: n?.value?.progress ?? 0,
            };

          case "E":
            return <RunnerEvent>{
              type: "runError",
              appOp,
              error: n.error,
            };

          case "C":
            return <RunnerEvent>{
              type: "runSuccess",
              appOp,
            };

          default:
            throw new Error("invalid notification of kind=" + n.kind);
        }
      })
    )
  );
};

type InlineInstallProgress = {
  globalProgress: number;
  itemProgress: number;
  currentAppOp: AppOp | null | undefined;
  installQueue: string[];
};

export const runAllWithProgress = (
  state: State,
  exec: Exec,
  precision = 100
): Observable<InlineInstallProgress> => {
  const total = state.uninstallQueue.length + state.installQueue.length;

  function globalProgress(s, localProgress) {
    let p =
      1 -
      (s.uninstallQueue.length + s.installQueue.length - localProgress) / total;
    p = Math.round(p * precision) / precision;
    return p;
  }

  return concat(
    ...getActionPlan(state).map((appOp) => runAppOp(state, appOp, exec))
  ).pipe(
    map((event) => {
      if (event.type === "runError") {
        throw event.error;
      }
      return <Action>{
        type: "onRunnerEvent",
        event,
      };
    }),
    scan(reducer, state),
    mergeMap((s) => {
      // Nb if you also want to expose the uninstall queue, feel free.
      const { currentProgressSubject, currentAppOp, installQueue } = s;

      if (!currentProgressSubject)
        return of({
          globalProgress: globalProgress(s, 0),
          itemProgress: 0,
          installQueue,
          currentAppOp,
        });

      // Expose more information about what's happening during the install
      return currentProgressSubject.pipe(
        map((v) => ({
          globalProgress: globalProgress(s, v),
          itemProgress: v,
          installQueue,
          currentAppOp,
        }))
      );
    }),
    distinctUntilChanged()
  );
};
// use for CLI, no change of the state over time
export const runAll = (state: State, exec: Exec): Observable<State> =>
  concat(
    ...getActionPlan(state).map((appOp) => runAppOp(state, appOp, exec))
  ).pipe(
    map(
      (event) =>
        <Action>{
          type: "onRunnerEvent",
          event,
        }
    ),
    reduce(reducer, state)
  );
export const runOneAppOp = (
  state: State,
  appOp: AppOp,
  exec: Exec
): Observable<State> =>
  runAppOp(state, appOp, exec).pipe(
    map(
      (event) =>
        <Action>{
          type: "onRunnerEvent",
          event,
        }
    ),
    reduce(reducer, state)
  );
export const runOne = (state: State, exec: Exec): Observable<State> => {
  const next = getNextAppOp(state);
  if (!next) return of(state);
  return runOneAppOp(state, next, exec);
};
