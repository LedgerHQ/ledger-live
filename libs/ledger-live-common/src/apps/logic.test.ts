// import { ManagerDeviceLockedError } from "@ledgerhq/errors";
import {
  initState,
  reducer,
  /*
  distribute,
  getActionPlan,
  predictOptimisticState,
  */
  updateAllProgress,
  getNextAppOp,
} from "./logic";
import { /*runAll,*/ runOneAppOp } from "./runner";
import {
  deviceInfo155,
  mockListAppsResult,
  mockExecWithInstalledContext,
} from "./mock";
// import { prettyActionPlan, prettyInstalled } from "./formatting";
import { setEnv } from "../env";
import { Action } from "./types";

setEnv("MANAGER_INSTALL_DELAY", 0);
// TODO reactivate this test after bitcoin 2.1.0 nano app

/*
const scenarios = [
  {
    name: "wipe installed apps",
    apps: "Bitcoin Legacy, Ethereum, Litecoin, Dogecoin, Ethereum Classic, XRP",
    installed: "Bitcoin Legacy, Litecoin, XRP, Ethereum Classic, Ethereum",
    actions: [
      {
        dispatch: {
          type: "wipe",
        },
        expectPlan:
          "-Litecoin, -Bitcoin Legacy, -XRP, -Ethereum Classic, -Ethereum",
        expectInstalled: "",
      },
    ],
  },
  {
    name: "install an app install its dep",
    apps: "Bitcoin Legacy, Litecoin, Dogecoin",
    installed: "",
    actions: [
      {
        dispatch: {
          type: "install",
          name: "Dogecoin",
        },
        expectPlan: "+Bitcoin Legacy, +Dogecoin",
        expectInstalled: "Bitcoin Legacy, Dogecoin",
      },
      {
        dispatch: {
          type: "install",
          name: "Litecoin",
        },
        expectPlan: "+Litecoin",
        expectInstalled: "Bitcoin Legacy, Dogecoin, Litecoin",
      },
    ],
  },
  {
    name: "install and uninstall an app",
    apps: "XRP",
    installed: "",
    actions: [
      {
        dispatch: {
          type: "install",
          name: "XRP",
        },
        expectPlan: "+XRP",
        expectInstalled: "XRP",
      },
      {
        dispatch: {
          type: "uninstall",
          name: "XRP",
        },
        expectPlan: "-XRP",
        expectInstalled: "",
      },
    ],
  },
  {
    name: "install an app install its dep",
    apps: "Bitcoin Legacy, Litecoin, Dogecoin",
    installed: "",
    actions: [
      {
        dispatch: {
          type: "install",
          name: "Dogecoin",
        },
        expectPlan: "+Bitcoin Legacy, +Dogecoin",
        expectInstalled: "Bitcoin Legacy, Dogecoin",
      },
      {
        dispatch: {
          type: "install",
          name: "Litecoin",
        },
        expectPlan: "+Litecoin",
        expectInstalled: "Bitcoin Legacy, Dogecoin, Litecoin",
      },
    ],
  },
  {
    name: "uninstall an app that have deps",
    apps: "Bitcoin Legacy, Litecoin, Dogecoin",
    installed: "Bitcoin Legacy, Litecoin, Dogecoin",
    actions: [
      {
        dispatch: {
          type: "uninstall",
          name: "Bitcoin Legacy",
        },
        expectPlan: "-Litecoin, -Dogecoin, -Bitcoin Legacy",
        expectInstalled: "",
      },
    ],
  },
  {
    name: "install existing is noop",
    apps: "Bitcoin Legacy",
    installed: "Bitcoin Legacy",
    actions: [
      {
        dispatch: {
          type: "install",
          name: "Bitcoin Legacy",
        },
        expectPlan: "",
        expectInstalled: "Bitcoin Legacy",
      },
    ],
  },
  {
    name: "uninstall non-existing is noop",
    apps: "Bitcoin Legacy",
    installed: "",
    actions: [
      {
        dispatch: {
          type: "uninstall",
          name: "Bitcoin Legacy",
        },
        expectPlan: "",
        expectInstalled: "",
      },
    ],
  },
  {
    name: "install an outdated app",
    apps: "Bitcoin Legacy",
    installed: "Bitcoin Legacy (outdated)",
    actions: [
      {
        dispatch: {
          type: "install",
          name: "Bitcoin Legacy",
        },
        expectPlan: "-Bitcoin Legacy, +Bitcoin Legacy",
        expectInstalled: "Bitcoin Legacy",
      },
    ],
  },
  {
    name: "install an outdated app dep",
    apps: "Bitcoin Legacy, Dogecoin",
    installed: "Bitcoin Legacy (outdated), Dogecoin (outdated)",
    actions: [
      {
        dispatch: {
          type: "install",
          name: "Dogecoin",
        },
        expectPlan: "-Dogecoin, -Bitcoin Legacy, +Bitcoin Legacy, +Dogecoin",
        expectInstalled: "Bitcoin Legacy, Dogecoin",
      },
    ],
  },
  {
    name: "install an app with outdated dep",
    apps: "Bitcoin Legacy, Litecoin, Dogecoin",
    installed: "Bitcoin Legacy (outdated), Litecoin (outdated)",
    actions: [
      {
        dispatch: {
          type: "install",
          name: "Dogecoin",
        },
        expectPlan:
          "-Litecoin, -Bitcoin Legacy, +Bitcoin Legacy, +Litecoin, +Dogecoin",
        expectInstalled: "Bitcoin Legacy, Litecoin, Dogecoin",
      },
    ],
  },
  {
    name: "install an outdated app with dependents",
    apps: "Bitcoin Legacy, Dogecoin",
    installed: "Bitcoin Legacy (outdated), Dogecoin (outdated)",
    actions: [
      {
        dispatch: {
          type: "install",
          name: "Bitcoin Legacy",
        },
        expectPlan: "-Dogecoin, -Bitcoin Legacy, +Bitcoin Legacy, +Dogecoin",
        expectInstalled: "Bitcoin Legacy, Dogecoin",
      },
    ],
  },
  {
    name: "update all will reinstall the outdated",
    apps: "Bitcoin Legacy, Litecoin, Ethereum",
    installed: "Bitcoin Legacy (outdated), Litecoin (outdated), Ethereum",
    actions: [
      {
        dispatch: {
          type: "updateAll",
        },
        expectPlan: "-Litecoin, -Bitcoin Legacy, +Bitcoin Legacy, +Litecoin",
        expectInstalled: "Ethereum, Bitcoin Legacy, Litecoin",
      },
    ],
  },
  {
    name: "update all still works with unknown apps",
    apps: "Bitcoin Legacy, Litecoin, Ethereum",
    installed:
      "Bitcoin Legacy (outdated), Litecoin (outdated), Ethereum, Unknown",
    actions: [
      {
        dispatch: {
          type: "updateAll",
        },
        expectPlan: "-Litecoin, -Bitcoin Legacy, +Bitcoin Legacy, +Litecoin",
        expectInstalled: "Ethereum, Unknown, Bitcoin Legacy, Litecoin",
      },
    ],
  },
  {
    name: "install and uninstall will undo (if top level dep)",
    apps: "Bitcoin Legacy",
    installed: "",
    actions: [
      {
        dispatch: [
          {
            type: "install",
            name: "Bitcoin Legacy",
          },
          {
            type: "uninstall",
            name: "Bitcoin Legacy",
          },
        ],
        expectPlan: "",
        expectInstalled: "",
      },
    ],
  },
  {
    name: "order is preserved in install action plan",
    apps: "Bitcoin Legacy, XRP, Ethereum, Ethereum Classic, Dogecoin, Zcash",
    installed: "",
    actions: [
      {
        dispatch: [
          {
            type: "install",
            name: "XRP",
          },
          {
            type: "install",
            name: "Ethereum Classic",
          },
          {
            type: "install",
            name: "Dogecoin",
          },
          {
            type: "install",
            name: "Zcash",
          },
        ],
        expectPlan:
          "+XRP, +Ethereum, +Ethereum Classic, +Bitcoin Legacy, +Dogecoin, +Zcash",
        expectInstalled:
          "XRP, Ethereum, Ethereum Classic, Bitcoin Legacy, Dogecoin, Zcash",
      },
    ],
  },
  {
    name: "order is preserved in uninstall action plan",
    apps: "Bitcoin Legacy, XRP, Ethereum, Ethereum Classic, Dogecoin, Zcash",
    installed:
      "XRP, Ethereum, Ethereum Classic, Bitcoin Legacy, Dogecoin, Zcash",
    actions: [
      {
        dispatch: [
          {
            type: "uninstall",
            name: "XRP",
          },
          {
            type: "uninstall",
            name: "Ethereum",
          },
          {
            type: "uninstall",
            name: "Dogecoin",
          },
          {
            type: "uninstall",
            name: "Bitcoin Legacy",
          },
        ],
        expectPlan:
          "-XRP, -Ethereum Classic, -Ethereum, -Dogecoin, -Bitcoin Legacy",
        expectInstalled: "",
      },
    ],
  },
];

scenarios.forEach((scenario) => {
  test("Scenario: " + scenario.name, async () => {
    let state = initState(
      mockListAppsResult(scenario.apps, scenario.installed, deviceInfo155)
    );
    expect(prettyActionPlan(getActionPlan(state))).toBe("");

    for (const action of scenario.actions) {
      state = (<any[]>[]).concat(action.dispatch).reduce(reducer, state);
      expect(prettyActionPlan(getActionPlan(state))).toBe(action.expectPlan);
      const optimisticState = predictOptimisticState(state);
      state = await runAll(
        state,
        mockExecWithInstalledContext(state.installed)
      ).toPromise();

      if (action.expectInstalled) {
        expect(prettyInstalled(state.installed)).toBe(action.expectInstalled);
      }
      state.currentProgressSubject = null;
      expect(state).toEqual(optimisticState);
      expect(prettyActionPlan(getActionPlan(state))).toBe("");
      const d: any = distribute(state);
      d.apps = d.apps.map(({ currency, ...rest }) => ({
        ...rest,
        currencyId: currency && currency.id,
      }));
      expect(d).toMatchSnapshot();
    }
  });
});

test("appsToRestore", async () => {
  const state = initState(
    mockListAppsResult(
      "Bitcoin Legacy, XRP, Ethereum, Ethereum Classic, Dogecoin, Zcash",
      "Bitcoin Legacy, Zcash",
      deviceInfo155
    ),
    ["Bitcoin Legacy", "XRP", "Dogecoin", "Zcash", "Ethereum Classic"]
  );
  expect(prettyActionPlan(getActionPlan(state))).toBe(
    "+XRP, +Dogecoin, +Ethereum, +Ethereum Classic"
  );
});
*/
/*
test("a lock error that occurs will not cancel the queue, another error will", () => {
  let state = initState(
    mockListAppsResult(
      "Bitcoin Legacy, XRP, Ethereum, Ethereum Classic, Dogecoin, Zcash",
      "",
      deviceInfo155
    )
  );

  state = [
    { type: "install", name: "Dogecoin" },
    { type: "install", name: "Ethereum Classic" }
  ].reduce(reducer, state);

  const plan = getActionPlan(state);

  state = reducer(state, {
    type: "onRunnerEvent",
    event: {
      type: "runError",
      appOp: plan[0],
      error: new ManagerDeviceLockedError()
    }
  });

  expect(getActionPlan(state)).toEqual(plan);
  expect(state.currentError).toEqual({
    appOp: plan[0],
    error: new ManagerDeviceLockedError()
  });

  state = reducer(state, { type: "recover" });
  expect(state.currentError).toBe(null);

  state = reducer(state, {
    type: "onRunnerEvent",
    event: {
      type: "runError",
      appOp: plan[0],
      error: new Error()
    }
  });

  expect(getActionPlan(state)).toEqual([]);
});
*/
test("global progress", async () => {
  let state = initState(
    mockListAppsResult(
      "Bitcoin Legacy, XRP, Ethereum, Ethereum Classic, Dogecoin, Zcash",
      "Bitcoin Legacy (outdated), Ethereum (outdated)",
      deviceInfo155
    )
  );
  expect(updateAllProgress(state)).toBe(1);
  state = [
    <Action>{
      type: "updateAll",
    },
  ].reduce(reducer, state);
  expect(updateAllProgress(state)).toBe(0);
  let next;
  let i = 0;
  const total = 4;

  while ((next = getNextAppOp(state))) {
    state = await runOneAppOp(
      state,
      next,
      mockExecWithInstalledContext(state.installed)
    ).toPromise();
    expect(updateAllProgress(state)).toBe(++i / total);
  }

  expect(i).toBe(total);
  expect(updateAllProgress(state)).toBe(1);
});
