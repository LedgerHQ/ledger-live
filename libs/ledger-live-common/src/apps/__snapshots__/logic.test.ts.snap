// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scenario: install an app install its dep 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "dogecoin",
      "name": "Dogecoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 49,
  "freeSpaceBytes": 200704,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 11,
  "totalAppsBytes": 45056,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install an app install its dep 2`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "dogecoin",
      "name": "Dogecoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "litecoin",
      "name": "Litecoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 48,
  "freeSpaceBytes": 196608,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 12,
  "totalAppsBytes": 49152,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install an app install its dep 3`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "dogecoin",
      "name": "Dogecoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 49,
  "freeSpaceBytes": 200704,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 11,
  "totalAppsBytes": 45056,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install an app install its dep 4`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "dogecoin",
      "name": "Dogecoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "litecoin",
      "name": "Litecoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 48,
  "freeSpaceBytes": 196608,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 12,
  "totalAppsBytes": 49152,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install an app with outdated dep 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "litecoin",
      "name": "Litecoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "dogecoin",
      "name": "Dogecoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 48,
  "freeSpaceBytes": 196608,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 12,
  "totalAppsBytes": 49152,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install an outdated app 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 50,
  "freeSpaceBytes": 204800,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 10,
  "totalAppsBytes": 40960,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install an outdated app dep 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "dogecoin",
      "name": "Dogecoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 49,
  "freeSpaceBytes": 200704,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 11,
  "totalAppsBytes": 45056,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install an outdated app with dependents 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "dogecoin",
      "name": "Dogecoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 49,
  "freeSpaceBytes": 200704,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 11,
  "totalAppsBytes": 45056,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install and uninstall an app 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "ripple",
      "name": "XRP",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 50,
  "freeSpaceBytes": 204800,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 10,
  "totalAppsBytes": 40960,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install and uninstall an app 2`] = `
Object {
  "apps": Array [],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 60,
  "freeSpaceBytes": 245760,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 0,
  "totalAppsBytes": 0,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install and uninstall will undo (if top level dep) 1`] = `
Object {
  "apps": Array [],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 60,
  "freeSpaceBytes": 245760,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 0,
  "totalAppsBytes": 0,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: install existing is noop 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 59,
  "freeSpaceBytes": 241664,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 1,
  "totalAppsBytes": 4096,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: order is preserved in install action plan 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "ripple",
      "name": "XRP",
    },
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "ethereum",
      "name": "Ethereum",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "ethereum_classic",
      "name": "Ethereum Classic",
    },
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "dogecoin",
      "name": "Dogecoin",
    },
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "zcash",
      "name": "Zcash",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 18,
  "freeSpaceBytes": 73728,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 42,
  "totalAppsBytes": 172032,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: order is preserved in uninstall action plan 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "zcash",
      "name": "Zcash",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 59,
  "freeSpaceBytes": 241664,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 1,
  "totalAppsBytes": 4096,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: uninstall an app that have deps 1`] = `
Object {
  "apps": Array [],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 60,
  "freeSpaceBytes": 245760,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 0,
  "totalAppsBytes": 0,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: uninstall non-existing is noop 1`] = `
Object {
  "apps": Array [],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 60,
  "freeSpaceBytes": 245760,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 0,
  "totalAppsBytes": 0,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: update all still works with unknown apps 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "ethereum",
      "name": "Ethereum",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": undefined,
      "name": "Unknown",
    },
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "litecoin",
      "name": "Litecoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 47,
  "freeSpaceBytes": 192512,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 13,
  "totalAppsBytes": 53248,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: update all will reinstall the outdated 1`] = `
Object {
  "apps": Array [
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "ethereum",
      "name": "Ethereum",
    },
    Object {
      "blocks": 10,
      "bytes": 40960,
      "currencyId": "bitcoin",
      "name": "Bitcoin",
    },
    Object {
      "blocks": 1,
      "bytes": 4096,
      "currencyId": "litecoin",
      "name": "Litecoin",
    },
  ],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 48,
  "freeSpaceBytes": 196608,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 12,
  "totalAppsBytes": 49152,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;

exports[`Scenario: wipe installed apps 1`] = `
Object {
  "apps": Array [],
  "appsSpaceBlocks": 60,
  "appsSpaceBytes": 245760,
  "freeSpaceBlocks": 60,
  "freeSpaceBytes": 245760,
  "osBlocks": 20,
  "osBytes": 81920,
  "shouldWarnMemory": false,
  "totalAppsBlocks": 0,
  "totalAppsBytes": 0,
  "totalBlocks": 80,
  "totalBytes": 327680,
}
`;
